### PR TITLE
add case for lifecycle with memory allocation

### PR DESCRIPTION
**Pull request by nanli1**
VIRT-297048: Life cycle with memory allocation
Signed-off-by: nanli <nanli@redhat.com>

```
 /usr/bin/python /usr/local/bin/avocado run --vt-type libvirt --test-runner=runner --vt-machine-type q35 memory.allocation.lifecycle

 (1/3) type_specific.io-github-autotest-libvirt.memory.allocation.lifecycle.positive_test.without_maxmemory: PASS (122.21 s)
 (2/3) type_specific.io-github-autotest-libvirt.memory.allocation.lifecycle.positive_test.with_numa: PASS (119.04 s)
 (3/3) type_specific.io-github-autotest-libvirt.memory.allocation.lifecycle.negative_test.with_maxmemory: PASS (7.00 s)


```

### CODE REVIEW

**Comment by nanli1** [comment]
modular_redesign

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        for item in xml_list:
+            libvirt_vmxml.check_guest_xml(vm_name, item)
+
+    def start_vm_and_login():
+        """
+        Start guest and login
+        """
+        if not vm.is_alive():
+            vm.start()
+        vm.wait_for_login().close()
```
Comment (liang-cong-red-hat) [inline libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:162]: Before the guest login of each life cycle step, we assume the vm must in running stateA, so if vm is not in running state, the case should fail.  If you change code, do not forget to add start vm code for managedsave.

Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:162]: Thanks lcong. Here are two points :

1. In the polarion case ,it's said ` 2) the guest login successfully` after reboot/resume.... , So I update it , which will only check login here with out ```
if not vm.is_alive():
            vm.start()  ```

2. For this sentence you said in comment :` Before the guest login of each life cycle step, we assume the vm must in running state, so if vm is not in running state,  the case should fail.  ` We cannot get it from the polarion case,If you want to check this point ,you may need to update polarion case and I will update auto code along with you

**Comment by nanli1** [comment]
@liang-cong-red-hat Updated

**Review by liang-cong-red-hat — APPROVED** [review]
LGTM

File: libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+                - without_maxmemory:
+                    xml_list = ["<memory unit='${mem_unit}'>${mem_value}</memory>", "<currentMemory unit='${current_mem_unit}'>${current_mem}</currentMemory>"]
+                    vm_attrs = {'memory_unit':'${mem_unit}','memory':${mem_value},'current_mem':${current_mem},'current_mem_unit':"${current_mem_unit}"}
+                - with_numa:
+                    max_mem = 15360000
+                    xml_list = ["<maxMemory slots='${max_mem_slots}' unit='${max_mem_unit}'>${max_mem}</maxMemory>","<memory unit='${mem_unit}'>${mem_value}</memory>", "<currentMemory unit='${current_mem_unit}'>${current_mem}</currentMemory>", "<cell id='0' cpus='0-1' memory='${numa_mem}'", "<cell id='1' cpus='2-3' memory='${numa_mem}'"]
+                    vm_attrs = {'max_mem_rt': ${max_mem}, 'max_mem_rt_slots': ${max_mem_slots}, 'max_mem_rt_unit': "${max_mem_unit}",'memory_unit':"${mem_unit}",'memory':${mem_value},'current_mem':${current_mem},'current_mem_unit':'${current_mem_unit}','vcpu': 4,'cpu': {'numa_cell': [{'id': '0', 'cpus': '0-1', 'memory': '${numa_mem}', 'unit': 'KiB'},{'id':'1','cpus': '2-3','memory':'${numa_mem}','unit':'KiB'}]}}
```
Comment (dzhengfy) [inline libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg:20]: It is too long for one parameter. Could you divide it into several parameters?

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+def run(test, params, env):
+    """
+    Verify memory allocation settings take effect during the life cycle of guest vm.
+
+    Scenario:
+    1:without max memory
+    2:with max memory
+    3:with numa topology
+    """
+    def setup_positive_test():
```
Comment (chunfuwen) [inline libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:23]: nothing actual logic happen here, please remove this method

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        """
+        Clean data.
+        """
+        test.log.info("TEST_TEARDOWN: Clean up env.")
+        if os.path.exists(save_file):
+            os.remove(save_file)
+        vm.undefine(options='--nvram --managed-save')
+        bkxml.sync()
+
+    def setup_negative_test():
```
Comment (chunfuwen) [inline libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:94]: see previous comment

File: libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    numa_mem = 1048576
+    variants case:
+        - positive_test:
+            variants mem_config:
+                - without_maxmemory:
+                    xml_list = ["<memory unit='${mem_unit}'>${mem_value}</memory>", "<currentMemory unit='${current_mem_unit}'>${current_mem}</currentMemory>"]
+                    vm_attrs = {'memory_unit':'${mem_unit}','memory':${mem_value},'current_mem':${current_mem},'current_mem_unit':"${current_mem_unit}"}
+                - with_numa:
+                    max_mem = 15360000
+                    xml_list = ["<maxMemory slots='${max_mem_slots}' unit='${max_mem_unit}'>${max_mem}</maxMemory>","<memory unit='${mem_unit}'>${mem_value}</memory>", "<currentMemory unit='${current_mem_unit}'>${current_mem}</currentMemory>", "<cell id='0' cpus='0-1' memory='${numa_mem}'", "<cell id='1' cpus='2-3' memory='${numa_mem}'"]
```
Comment (dzhengfy) [inline libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg:19]: It is not suitable to match the hardcoded xml. 
Could you help use below pr to match your desired xml?
- https://github.com/avocado-framework/avocado-vt/pull/3675

File: libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+                    vm_attrs = {'max_mem_rt': ${max_mem}, 'max_mem_rt_slots': ${max_mem_slots}, 'max_mem_rt_unit': "${max_mem_unit}",'memory_unit':"${mem_unit}",'memory':${mem_value},'current_mem':${current_mem},'current_mem_unit':'${current_mem_unit}','vcpu': 4,'cpu': {'numa_cell': [{'id': '0', 'cpus': '0-1', 'memory': '${numa_mem}', 'unit': 'KiB'},{'id':'1','cpus': '2-3','memory':'${numa_mem}','unit':'KiB'}]}}
+        - negative_test:
+            variants mem_config:
+                - with_maxmemory:
+                    vm_attrs = {'max_mem_rt': ${max_mem}, 'max_mem_rt_slots': ${max_mem_slots}, 'max_mem_rt_unit': '${max_mem_unit}','memory_unit':'${mem_unit}','memory':${mem_value},'current_mem':${current_mem},'current_mem_unit':'${current_mem_unit}'}
```
Comment (dzhengfy) [inline libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg:24]: Please divide it into several parameters to improve the readibility.

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        test.log.info("TEST_STEP4:Check vm suspend & resume and memory config")
+        virsh.suspend(vm_name, ignore_status=False, debug=True)
+        virsh.resume(vm_name, ignore_status=False, debug=True)
+        vm.wait_for_login().close()
+        check_mem_config()
+
+        test.log.info("TEST_STEP5:Check vm save & restore and memory config")
+        if os.path.exists(save_file):
+            os.remove(save_file)
+        virsh.save(vm_name, save_file, ignore_status=False)
```
Comment (dzhengfy) [inline libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:64]: Do you need `debug=True`?

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        virsh.restore(save_file, ignore_status=False)
```
Comment (dzhengfy) [inline libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:65]: `debug=True`?

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        pass
+
+    def run_negative_test():
+        """
+        Define vm and check result.
+        """
+        test.log.info("TEST_STEP1: Define vm with maxMemory and no numa")
+        vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+        vmxml.setup_attrs(**vm_attrs)
+        test.log.debug("Define vm with %s." % vmxml)
```
Comment (dzhengfy) [inline libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:107]: It should be `'test.log.debug("Define vm with %s.", vmxml)'`

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    xml_list = eval(params.get("xml_list", '{}'))
+    vm_attrs = eval(params.get("vm_attrs"))
+    save_file = "/tmp/%s.save" % vm_name
+
+    run_test = eval("run_%s" % case)
+    setup_test = eval("setup_%s" % case)
+    teardown_test = eval("teardown_%s" % case)
+
+    try:
+        setup_test()
```
Comment (dzhengfy) [inline libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:182]: Currently there is no setup steps for these cases, so it is ok to remove it. We can add it in future if the cases need it.

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py (hunk 3 above)
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py]: Thanks chunfu , done

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py (hunk 4 above)
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py]: Thanks chunfu , done

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        libvirt.check_result(cmd_result, error_msg)
+
+    def teardown_negative_test():
+        """
+        Clean data.
+        """
+        test.log.info("TEST_TEARDOWN: Clean up env.")
+        bkxml.sync()
+
+    def check_qemu_cmdline():
```
[the remark this answers is not in the document]
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py:108]: Yes ,Thank you , done

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        if dominfo_dict["Used memory"] != str(current_mem):
+            test.fail("Current memory should be %s instead of %s " % (
+                current_mem, dominfo_dict["Used memory"]))
+
+    def check_mem_config():
+        """
+        Check current mem device config
+        """
+        for item in xml_list:
+            libvirt_vmxml.check_guest_xml(vm_name, item)
```
[the remark this answers is not in the document]
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py]: Yest done

File: libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg (hunk 2 above)
Comment (nanli1) [reply libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg]: done

File: libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg (hunk 5 above)
Comment (nanli1) [reply libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg]: done

File: libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg (hunk 6 above)
Comment (nanli1) [reply libvirt/tests/cfg/mem/memory_allocation/lifecycle_with_memory_allocation.cfg]: done

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py (hunk 7 above)
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py]: done

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py (hunk 8 above)
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py]: done

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py (hunk 9 above)
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py]: done

File: libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py (hunk 10 above)
Comment (nanli1) [reply libvirt/tests/src/mem/memory_allocation/lifecycle_with_memory_allocation.py]: done

**Review by chunfuwen — APPROVED** [review]
lgtm

**Review by dzhengfy — APPROVED** [review]

**Pull request merged**